tests/test_remapper.py

```diff
@@ -1,2 +1,124 @@
+import pytest
+import sparse
+import xarray as xr
+
+from xgriddedaxis import Remapper
+from xgriddedaxis.remapper import _INCOMING_KEY, _OUTGOING_KEY
+
+from .utils import create_dataset, generate_time_and_bounds
+
+
+@pytest.fixture(scope='module')
+def incoming():
+    n = 13
+    bounds = np.round(np.logspace(2.0, 3.5, num=n), decimals=0)
+    fractions = np.round(np.random.random(n - 1), decimals=1)
+    incoming = generate_time_and_bounds(bounds, fractions)
+    return incoming
+
+
+@pytest.fixture(scope='module')
+def outgoing():
+    n = 25
+    bounds = np.round(np.logspace(2.0, 3.6, num=n), decimals=0)
+    fractions = np.round(np.random.random(n - 1), decimals=1)
+    outgoing = generate_time_and_bounds(bounds, fractions)
+    return outgoing
+
+
+@pytest.fixture(scope='module')
+def dataset(incoming, outgoing):
+    ds = create_dataset(incoming['time'], incoming['time_bounds'])
+    return ds
+
+
+def test_remapper_init(incoming, outgoing):
+    remapper = Remapper(
+        incoming_axis=incoming,
+        outgoing_axis=outgoing,
+        axis_name='time',
+        boundary_variable='time_bounds',
+    )
+    assert isinstance(remapper.weights.data, sparse._coo.core.COO)
+    assert remapper.weights.shape == (
+        remapper.outgoing_axis[remapper.axis_name].size,
+        remapper.incoming_axis[remapper.axis_name].size,
+    )
+    assert set(remapper.weights.dims) == set([_OUTGOING_KEY, _INCOMING_KEY])
+
+
+def test_remapper_apply_weights(dataset, incoming, outgoing):
+    remapper = Remapper(
+        incoming_axis=incoming,
+        outgoing_axis=outgoing,
+        axis_name='time',
+        boundary_variable='time_bounds',
+    )
+    remapped_data = remapper(dataset.x)
+    assert remapped_data.shape == (24, 2, 2)
+    assert set(remapped_data.dims) == set(dataset.x.dims)
+
+
+def test_remapper_apply_weights_dask(dataset, incoming, outgoing):
+    remapper = Remapper(
+        incoming_axis=incoming,
+        outgoing_axis=outgoing,
+        axis_name='time',
+        boundary_variable='time_bounds',
+    )
+    ds = dataset.chunk()
+    remapped_data = remapper(ds.x)
+    assert isinstance(remapped_data.data, dask.array.Array)
+    assert remapped_data.shape == (24, 2, 2)
+    assert set(remapped_data.dims) == set(dataset.x.dims)
+    xr.testing.assert_equal(remapped_data, remapper(dataset.x))
+
+
+def test_remapper_apply_weights_invalid_input(dataset, incoming, outgoing):
+    remapper = Remapper(
+        incoming_axis=incoming,
+        outgoing_axis=outgoing,
+        axis_name='time',
+        boundary_variable='time_bounds',
+    )
+
+    with pytest.raises(NotImplementedError):
+        _ = remapper(dataset)
+
+    with pytest.raises(TypeError):
+        _ = remapper(dataset.x.data)
+
+
+def test_bounds_sanity_check():
+    from xgriddedaxis.remapper import _bounds_sanity_check
+
+    bounds = np.array([0.0, 15.0, 10.0, 22.0, 26.0, 50.0, 45.0])
+    fractions = np.array([0.5, 0.4, 0.2, 1, 0.5, 0.3])
```

Comment:
Looking at this, it makes me realize that there are a couple other things to check:

1. Two elements are the same:

    ```python
    bounds = np.array([0.0, 1.0, 2.0])
    fractions = np.array([1, 0])
    ```

    This should produce a `time` coordinate with:

    ```python
    time = np.array([1.0, 1.0])
    ```

    Which should trigger an error, since it is not monotonic.

2. Negative fractions?  Is this allowable?  How do we check for them?  What about `fractions` greater than 1?

Reply:
Thank you for catching these two. 

1) Should be fixed now.

2) I am looking into it

Reply:
@kmpaul, 

Regarding negative fractions, I think we should throw an error.

Say we have the following:

```python
In [32]: fractions                                                                                 
Out[32]: array([-0.96, -0.84, -0.48, -0.3 ])

In [33]: bounds                                                                                    
Out[33]: array([ 100.,  251.,  631., 1585., 3981.])
```

This generates:

```python
time = array([-44.96, -68.2 , 173.08, 866.2 ])
time_bounds = array([[ 100.,  251.],
       [ 251.,  631.],
       [ 631., 1585.],
       [1585., 3981.]])
```

I am envisioning a sanity check that loops over both arrays, and asserts that each data tick fits into its bounds i.e.

```python
In [39]: for val, bds in zip(time, time_bounds): 
    ...:     assert np.logical_and(bds[0]<=val, val<=bds[1]) 
```

What do you think?

Reply:
Regarding `fractions` greater than `1`, the assertion above takes of this case as well

```python
In [43]: fractions = np.array([1.96, 0.84, 0.48, 0.3])                                             

In [44]: bounds                                                                                    
Out[44]: array([ 100.,  251.,  631., 1585., 3981.])

In [48]: time                                                                                      
Out[48]: array([ 395.96,  570.2 , 1088.92, 2303.8 ])

In [49]: time_bounds                                                                               
Out[49]: 
array([[ 100.,  251.],
       [ 251.,  631.],
       [ 631., 1585.],
       [1585., 3981.]])

In [50]: for val, bds in zip(time, time_bounds): 
    ...:     print(val, bds) 
    ...:     assert np.logical_and(bds[0]<=val, val<=bds[1]) 
    ...:      
    ...:                                                                                           
395.96 [100. 251.]
``